### ISSUE TITLE
Inconsistent inference inside print()

### ISSUE THREAD

**Issue by syimyuzya** [python/mypy#6017]
Just found that mypy infers the type of `sum(l)` in the code below differently depending on whether it is inside a `print()`:

```python
#! /usr/bin/python3

# Run this with `mypy --disallow-any-expr --show-column-numbers`

from typing import List

l: List[complex] = [1.0, 2.0, 3.0-2.0j]

reveal_type(sum(l))  # builtins.complex*

s = sum(l)
reveal_type(s)  # builtins.complex*
print(s)  # typechecks!

print(sum(l))  # but this line does not typecheck!
# 15:7: error: Expression type contains "Any" (has type "Union[Any, int]")
```

Tested with mypy 0.641 and 0.650+dev.ad2d4ba1db9699779d5705094f991d3bb3ac112d (current master branch) on Python 3.7.1, results are same.

**Comment by ilevkivskyi** [comment]
Yeah, `--disallow-any-expr` is known to be problematic, we have received several reports about it recently.

**Comment by Levitanus** [comment]
Append to the false-positive stack, please:
```Python
import typing as ty


class ChildAddress(ty.Tuple[int, int]):
    """Represents MIDI (Bus, Channel) of child Track.

    Can be compared to tuple. All (busses/channels) are equal to one.
    """

    def __new__(cls, bus: int, channel: int) -> 'ChildAddress':
        return super().__new__(cls, (bus, channel))
        # Argument 2 to "__new__" of "tuple" has incompatible type "Tuple[int, int]"; expected "Iterable[_T_co]"

    def __init__(self, bus: int, channel: int) -> None:
        """make immutable address.

        Parameters
        ----------
        bus : int
        channel : int
        """
        super().__init__()

    @property
    def bus(self) -> int:
        """MIDI Bus.

        Returns
        -------
        int
            0 if All buses
            -1 if no midi routing
        """
        return self[0]

    @property
    def channel(self) -> int:
        """Midi Channel.

        Returns
        -------
        int
            0 if All channels
            -1 if no midi routing
        """
        return self[1]

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, ty.Sequence):
            return False
        if len(other) != 2:
            return False
        for item in other:
            if not isinstance(item, int):
                return False
            if item > 16:
                return False
        other = ty.cast(ty.Tuple[int, int], other)
        if other[0] != self[0]:
            for item in ty.cast(ty.Tuple[int, int], (other[0], self[0])):
                #  Redundant cast to "Tuple[int, int]"
                if item != '0':  # Expression has type "Any"
                    return False
                else:
                    break
        if other[1] != self[1]:
            for item in (other[1], self[1]):
                if item != 0:  # Expression has type "Any"
                    return False
                else:
                    break
        return True

    def __repr__(self) -> str:
        return f'ChildAddress(bus={self.bus}, channel={self.channel})'

    def __hash__(self) -> int:
        return hash((self[0], self[1]))

```

**Comment by AlexWaygood** [comment]
The false-positive error is [no longer emitted](https://mypy-play.net/?mypy=0.700&python=3.10&flags=disallow-any-expr&gist=98ed24dba557e879548bac8702d73ad2) on mypy 0.700+
